internal/util/prometheus.go

```diff
@@ -0,0 +1,55 @@
+package util
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+)
+
+type ControllerFunctionalPrometheusMetrics struct {
+	// number of post /config to proxy successfully
+	ConfigPassCounter prometheus.Counter
+	// number of post /config to proxy failed
+	ConfigFailureCounter prometheus.Counter
+	// number of ingress analysis failure
+	ParseFailureCounter prometheus.Counter
+
+	// duration of last successful confiuration sync
+	ConfigureDurationHistogram prometheus.Histogram
+}
+
+func (ctrlMetrics *ControllerFunctionalPrometheusMetrics) GeneratePrometheusCounter(name, help string) prometheus.Counter {
```

Comment:
On methods, it's idiomatic that receiver names should be as short as possible. Prefer `m` or `cm` over `ctrlMetrics`.

Reply:
leave as it is.

Reply:
There are at least two reasons to name the receiver `c` or `cm`:
- The official code review suggestions for Go on [receiver naming](https://github.com/golang/go/wiki/CodeReviewComments#receiver-names)
- Consistency with the rest of our codebase. We use such 1- or 2-letter abbreviations for receiver names everywhere I can think of.

Therefore, if there's no compelling reason to break consistency with the above, please prefer sticking to our conventions.
If there is a compelling reason, let's discuss it here.

Reply:
@mflendrich do you mind do a "Commit Request" review here so that I understand what exact function receiver name in your mind ? 🙇

Reply:
@ccfishk what's desired here is that a change like this be made:

```diff
- func (ctrlMetrics *ControllerFunctionalPrometheusMetrics) GeneratePrometheusCounter(name, help string) prometheus.Counter {
+ func (cm *ControllerFunctionalPrometheusMetrics) GeneratePrometheusCounter(name, help string) prometheus.Counter {
```

This is more common (e.g. what he meant in this case by "idiomatic" to some extent) and is more consistent with the rest of the codebase.

Reply:
not an issue anymore.
https://github.com/Kong/kubernetes-ingress-controller/blob/d0d14553f8bacff143d3319560b102f080d1022a/internal/util/prometheus.go#L20